### PR TITLE
handle neg expiry

### PR DESCRIPTION
**Pull request by jennyf19**
#1419

### CODE REVIEW

File: src/Microsoft.Identity.Web/TokenCacheProviders/InMemory/MsalMemoryTokenCacheProvider.cs [hunk 1]
```diff
@@ -92,6 +92,10 @@ protected override Task WriteCacheBytesAsync(string cacheKey, byte[] bytes)
             if (cacheSerializerHints != null && cacheSerializerHints?.SuggestedCacheExpiry != null)
             {
                 cacheExpiry = cacheSerializerHints.SuggestedCacheExpiry.Value.UtcDateTime - DateTime.UtcNow;
+                if (cacheExpiry < TimeSpan.Zero)
+                {
+                    cacheExpiry = TimeSpan.FromMilliseconds(1);
+                }
```
Comment (jmprieur) [inline src/Microsoft.Identity.Web/TokenCacheProviders/InMemory/MsalMemoryTokenCacheProvider.cs:98]: Don't we want to add logging here too?

**Review by jmprieur — APPROVED** [review]
LGTM
Left 2 suggestions
Thanks @jennyf19

File: src/Microsoft.Identity.Web/TokenCacheProviders/InMemory/MsalMemoryTokenCacheProvider.cs (hunk 1 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web/TokenCacheProviders/InMemory/MsalMemoryTokenCacheProvider.cs:98]: we don't already have it here, so didn't want to add it just for this.

File: src/Microsoft.Identity.Web/TokenCacheProviders/Distributed/MsalDistributedTokenCacheAdapter.Logger.cs [hunk 2]
```diff
@@ -58,6 +58,12 @@ private static class Logger
                     LoggingEventId.MemoryCacheRead,
                     "[MsIdWeb] {CacheType}: {Operation} cacheKey {CacheKey} cache size {Size} ");
 
+            private static readonly Action<ILogger, string, string, Exception?> s_l1CacheNegativeExpiry =
+                LoggerMessage.Define<string, string>(
+                    LogLevel.Debug,
+                    LoggingEventId.MemoryCacheNegativeExpiry,
+                    "[MsIdWeb] {CacheType}: {Operation} The SuggestedCacheExpiry from MSAL was negative. ");
```
[the remark this answers is not in the document]
Comment (jennyf19) [reply src/Microsoft.Identity.Web/TokenCacheProviders/Distributed/MsalDistributedTokenCacheAdapter.Logger.cs]: fixed. thanks

[the remark this answers is not in the document]
Comment (jmprieur) [reply src/Microsoft.Identity.Web/TokenCacheProviders/Distributed/MsalDistributedTokenCacheAdapter.Logger.cs]: :shipit:

File: src/Microsoft.Identity.Web/TokenCacheProviders/Distributed/MsalDistributedTokenCacheAdapter.Logger.cs [hunk 3]
```diff
@@ -58,6 +58,12 @@ private static class Logger
                     LoggingEventId.MemoryCacheRead,
                     "[MsIdWeb] {CacheType}: {Operation} cacheKey {CacheKey} cache size {Size} ");
 
+            private static readonly Action<ILogger, string, string, Exception?> s_l1CacheNegativeExpiry =
+                LoggerMessage.Define<string, string>(
+                    LogLevel.Debug,
```
Comment (bgavrilMS) [inline src/Microsoft.Identity.Web/TokenCacheProviders/Distributed/MsalDistributedTokenCacheAdapter.Logger.cs:63]: More like a LogLevel.Warning or Error?

Comment (jennyf19) [reply src/Microsoft.Identity.Web/TokenCacheProviders/Distributed/MsalDistributedTokenCacheAdapter.Logger.cs:63]: i don't think this error is actionable by the service.

Comment (bgavrilMS) [reply src/Microsoft.Identity.Web/TokenCacheProviders/Distributed/MsalDistributedTokenCacheAdapter.Logger.cs:63]: They should log a bug on MSAL

Comment (jennyf19) [reply src/Microsoft.Identity.Web/TokenCacheProviders/Distributed/MsalDistributedTokenCacheAdapter.Logger.cs:63]: i did.

File: src/Microsoft.Identity.Web/TokenCacheProviders/Distributed/MsalDistributedTokenCacheAdapter.cs [hunk 4]
```diff
@@ -237,6 +237,11 @@ protected override async Task WriteCacheBytesAsync(string cacheKey, byte[] bytes
             if (cacheSerializerHints != null && cacheSerializerHints?.SuggestedCacheExpiry != null)
             {
                 cacheExpiry = cacheSerializerHints.SuggestedCacheExpiry.Value.UtcDateTime - DateTime.UtcNow;
+                if (cacheExpiry < TimeSpan.Zero)
```
Comment (bgavrilMS) [inline src/Microsoft.Identity.Web/TokenCacheProviders/Distributed/MsalDistributedTokenCacheAdapter.cs:240]: Hmm, should you not delete the cache entry instead?

Comment (jennyf19) [reply src/Microsoft.Identity.Web/TokenCacheProviders/Distributed/MsalDistributedTokenCacheAdapter.cs:240]: we could but then the L1/L2 will be out of sync.

File: tests/Microsoft.Identity.Web.Test/L1L2CacheTests.cs [hunk 5]
```diff
@@ -51,6 +52,30 @@ public async Task WriteCache_WritesInL1L2_TestAsync()
...
+            CacheSerializerHints cacheSerializerHints = new CacheSerializerHints();
+            cacheSerializerHints.SuggestedCacheExpiry = System.DateTimeOffset.Now - System.TimeSpan.FromHours(1);
+
+            // Act
+            await _testCacheAdapter.TestWriteCacheBytesAsync(DefaultCacheKey, cache, cacheSerializerHints).ConfigureAwait(false);
+
+            // Assert
+            Assert.Equal(1, _testCacheAdapter._memoryCache.Count);
+            Assert.Single(L2Cache.dict);
+            await Task.Delay(1000).ConfigureAwait(false);
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Web.Test/L1L2CacheTests.cs:72]: Generally we should avoid awaiting in tests. But I think in this case the adapter should just delete the entry itself or not set an expiry.

File: tests/Microsoft.Identity.Web.Test/L1L2CacheTests.cs [hunk 6]
```diff
@@ -51,6 +52,30 @@ public async Task WriteCache_WritesInL1L2_TestAsync()
             Assert.Single(L2Cache.dict);
         }
 
+        [Fact]
+        public async Task WriteCache_NegativeExpiry_TestAsync()
+        {
+            // Arrange
+            byte[] cache = new byte[3];
+            AssertCacheValues(_testCacheAdapter);
+            Assert.Equal(0, _testCacheAdapter._memoryCache.Count);
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Web.Test/L1L2CacheTests.cs:61]: nit: internal member naming is MemoryCache, not _memoryCache

File: tests/Microsoft.Identity.Web.Test/L1L2CacheTests.cs (hunk 5 above)
Comment (jennyf19) [reply tests/Microsoft.Identity.Web.Test/L1L2CacheTests.cs:72]: what happens if MSAL expects that the entry was written and doesn't find it just after? I tried to have a void implementation of a serializer and that didn't work. MSAL crashed, it was expecting something to come back.

**Review by bgavrilMS — APPROVED** [review]
LGTM, I think we should still log an error or a warning, to try to expose this problem somehow and get MSAL fixed.

**Pull request merged**